zkevm-circuits/src/evm_circuit/op_execution/dup.rs

```diff
@@ -0,0 +1,462 @@
+                    - (state_curr.program_counter.expr() + 1.expr()),
+                // dupx contains one time push operation
+                state_next.stack_pointer.expr()
+                    - (state_curr.stack_pointer.expr() - 1.expr()),
+                state_next.gas_counter.expr()
+                    - (state_curr.gas_counter.expr() + GasCost::FASTEST.expr()),
+            ];
+
+            let DupSuccessAllocation {
+                case_selector,
+                word,
+                selectors,
+            } = &self.success;
+
+            let mut dup_constraints = vec![];
+            for idx in 0..16 {
+                // selector can transit from 1 to 0 only once as [1, 1, 1, ...,
+                // 0, 0, 0]
+                if idx > 0 {
+                    let diff =
+                        selectors[idx - 1].expr() - selectors[idx].expr();
+                    dup_constraints.push(diff.clone() * (1.expr() - diff));
+                }
+                // selectors needs to be 0 or 1
+                dup_constraints.push(
+                    selectors[idx].expr() * (1.expr() - selectors[idx].expr()),
+                );
+            }
+
+            let selectors_sum =
+                selectors.iter().fold(0.expr(), |sum, s| sum + s.expr());
+            dup_constraints.push(selectors_sum - num_duplicated.clone());
+
+            let bus_mapping_lookups = [
+                // TODO: add 32 Bytecode lookups when supported
+                vec![Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: num_duplicated.clone() - 1.expr(),
+                    value: word.expr(),
+                    is_write: false,
+                })],
+                vec![Lookup::BusMappingLookup(BusMappingLookup::Stack {
+                    index_offset: (-1).expr(), // fixed as DUP decreases the stack pointer
+                    value: word.expr(),
+                    is_write: true,
+                })],
+            ]
+            .concat();
+
+            Constraint {
+                name: "DupGadget success",
+                selector: case_selector.expr(),
+                polys: [state_transition_constraints, dup_constraints].concat(),
+                lookups: bus_mapping_lookups, //vec![]
+            }
+        };
+
+        let stack_overflow = {
+            let stack_pointer = state_curr.stack_pointer.expr();
+            Constraint {
+                name: "DupGadget stack overflow",
+                selector: self.stack_overflow.expr(),
+                polys: vec![stack_pointer],
+                lookups: vec![],
+            }
+        };
+
+        let stack_pointer = state_curr.stack_pointer.expr();
+        let diff = 1024.expr() - stack_pointer - num_duplicated;
+
+        // diff's maxium is 1023 when stack_pointer = 0 and position = 1,
+        // the minimum is 0 when stack_pointer + position = 1024 , i.e. stack_point = 1020, DUPX = 4
+        let stack_underflow = {
```

Comment:
Why is maximum `1023`? The underflow case should be `0 <= num_duplicated + stack_pointer - 1025 < 16`, which fits in a `Range16` lookup.